### PR TITLE
[wip] TFE Keras integration

### PR DESCRIPTION
**Pull request by mortendahl**
- [x] new release of TFE
- [ ] tests for `tfe.py` to increase coverage (see [CONTRIBUTING](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md))
- [x] last run through notebooks
- [x] rename notebook directory?

### CODE REVIEW

File: syft/frameworks/keras/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from . import model, layers
+from syft.frameworks.keras.hook import KerasHook
+
+
+def get_hooks(torch=None, keras=None):
+    # TODO[jason]: this is currently very ugly, in preparation for merging keras with existing workers
+    #              for now, we ignore this in favor of TFWorker; sync on this with the team
```
[the remark this answers is not in the document]
Comment (robert-wagner) [reply syft/frameworks/keras/__init__.py]: Still really don't like the terms client and server worker. Any ideas on this front?

[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/__init__.py]: client vs worker? we chose these to mirror the existing terminology, so could also be beyond the scope of this PR

File: syft/frameworks/keras/hook.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+class KerasHook:
+    def __init__(self, keras):
+        self.keras = keras
+        self._hook_layers()
+        self._hook_sequential()
+
+    def _hook_layers(self):
+        for layer_cls in filter_nonlayers(self.keras.layers, tfe.keras.layers):
+            registered_cls = add_constructor_registration(layer_cls)
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/hook.py]: nope! leftover from a previous version, will fix

File: syft/frameworks/keras/model/sequential.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        # Default is SecureNN
+        prot = 'securenn'
+    if isinstance(prot, str):
+        # here lies the dumbest, sneakiest bug of my life, so far...
+        # if you move this constructor call into _protocol_map,
+        # the protocol objects will be initialized before set_config above is called,
+        # which means any direct calls to prot will use the default device names
+        # (which come from LocalConfig). so we return the class from _protocol_map
+        # and instantiate it here
+        prot = _protocol_map[prot]()
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/model/sequential.py]: when the workers are specified, the config is set just above, and so tfe will construct the protocol accordingly.  when the workers are not specified, we assume the user has created their own config outside of the scope of the `share` method. this seemed to me to be most flexible behavior for the user while conforming more closely to the `.share()` signature.  do you think there's a better way?

File: syft/frameworks/keras/model/sequential.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+
+def _rebuild_tfe_model(keras_model, stored_keras_weights):
+    tfe_model = tfe.keras.Sequential()
+
+    for keras_layer in keras_model.layers:
+        tfe_layer = _instantiate_tfe_layer(keras_layer, stored_keras_weights)
+        tfe_model.add(tfe_layer)
+
+        if hasattr(tfe_layer, '_batch_input_shape'):
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/model/sequential.py]: 'batch_input_shape` or `input_shape` must be provided as a kwarg to the first layer in a Sequential model.  usually it's `input_shape` and the batch dim is None, but here we stick to `batch_input_shape` due to https://github.com/tf-encrypted/tf-encrypted/issues/499#issuecomment-495412270

File: syft/workers/tensorflow_server.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+"""To be extended in the near future."""
+
+class TensorflowServerWorker:
+    def __init__(self, host):
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/workers/tensorflow_server.py]: this would likely change the semantics of `model.share()`, so I'm not sure it's worth it unless we can figure out a way around that

File: syft/workers/tensorflow_client.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class TensorflowClientWorker:
+    def connect_to_model(self, config_filename, input_shape, output_shape):
+
+        config = tfe.RemoteConfig.load(config_filename)
+        tfe.set_config(config)
+
+        # We set the protocol to Pond. The protocol
+        # selected shouldn't matter on the client side
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/workers/tensorflow_client.py]: we should rethink this with the rest of the config questions (cc @yanndupis)

File: syft/workers/tfe.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    # TODO(Morten) this should be turned into a proxy, with existing code
+    # extracted into a new component that's launched via a script
+
+    def __init__(self, host=None):
+        self.host = host
+        self._server_process = None
+
+    def start(self, player_name, *workers):
+
+        config_filename = "/tmp/tfe.config"
```
Comment (yanndupis) [inline syft/workers/tfe.py]: What's the benefit of saving this config file now?

File: syft/frameworks/keras/model/sequential.py [hunk 8]
```diff
@@ -79,58 +75,35 @@ def step_fn():
...
-
-
-def _launch_tfserver_subprocesses(model):
-    subprocess_calls = {}
-    cmd = "python -m tf_encrypted.player --config /tmp/tfe.config {}"
-    hostmap = tfe.get_config().hostmap
-    for player_name in hostmap.keys():
-        subprocess_calls[player_name] = subprocess.Popen(cmd.format(player_name).split(' '))
-    setattr(model, '_subprocess_calls', subprocess_calls)
+    for worker in model._workers:
```
Comment (yanndupis) [inline syft/frameworks/keras/model/sequential.py]: Do we want call it `shutdown` or `workers_shutdown`? Because the workers launched is abstracted in share, it's not very clear what's the intent of shutdown.

File: examples/experimental/tfe_integration_demo/Step1_Public_Training.ipynb [hunk 9]
```diff
@@ -0,0 +1,185 @@
+{
+ "cells": [
```
Comment (mari-linhares) [inline examples/experimental/tfe_integration_demo/Step1_Public_Training.ipynb]: I think this notebook could be merged with step 2.

File: syft/frameworks/keras/model/sequential.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    'kernel_constraint',
+    'bias_constraint',
+    'dilation_rate',
+]
+
+
+def share(model, *workers, target_graph=None, init_run_tag=None):
+
+    # Store Keras weights before loading them in the TFE layers.
+    # TODO(Morten) we could optimize runtime by running a single model.get_weights instead
```
[the remark this answers is not in the document]
Comment (yanndupis) [reply syft/frameworks/keras/model/sequential.py]: I did this way in order to load the right weights from the `tf.keras.layer` into the `tfe.keras.layer` based on the layer name.  The transition between `tf.keras.layer` to `tfe.keras.layer`is happening in ` _instantiate_tfe_layer`. Also in tfe currently we don't have a `mode.set_weights()`to load the weights from the entire model into a tfe model. Maybe there a better way to do this :)

File: syft/frameworks/keras/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/__init__.py]: this should be outdated -- we now have a single TFEWorker, the comment needs to be updated

File: syft/frameworks/keras/model/sequential.py (hunk 8 above)
Comment (jvmancuso) [reply syft/frameworks/keras/model/sequential.py]: yup, `shutdown_workers()` seems appropriate

File: syft/__init__.py [hunk 11]
```diff
@@ -49,6 +50,7 @@
     "LoggingTensor",
     "PointerTensor",
     "VirtualGrid",
+    "keras",
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/__init__.py]: we no longer do, this was an artifact of a previous commit that I missed during cleanup

File: syft/frameworks/keras/model/sequential.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/model/sequential.py]: I think this makes the code neater, and also the optimization we'd get is negligible in the context of the rest of the function

File: syft/frameworks/keras/model/sequential.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    for worker in model._workers:
+        worker.stop()
+
+
+def _configure_tfe(workers):
+
+    if not workers or len(workers) != 3:
+        raise RuntimeError("TF Encrypted expects three parties for its sharing protocols.")
+
+    player_to_worker_mapping = OrderedDict()
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/model/sequential.py]: these are player names, not quite the same as worker names. in TFE, a player represents a specific role in the MPC game, which means they're generally not symmetric.  in theory we could disambiguate worker names and these player names, but currently TFE expects specific names in order to distribute computation according to a given protocol, and that seems like unnecessary complexity for this first go

[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/keras/model/sequential.py]: but yes, the method _config_from_workers can be converted to a classmethod and used here to avoid code duplication

File: examples/experimental/tfe_integration_demo/Step1_Public_Training.ipynb (hunk 9 above)
Comment (jvmancuso) [reply examples/experimental/tfe_integration_demo/Step1_Public_Training.ipynb]: agreed, there's an incoming commit that fixes this flow

File: examples/experimental/tfe_integration_demo/step2_tfe_private_prediction.ipynb [hunk 13]
```diff
@@ -0,0 +1,274 @@
+{
```
Comment (yanndupis) [inline examples/experimental/tfe_integration_demo/step2_tfe_private_prediction.ipynb]: I think we might want to remove this notebook, so we present only one approach (server-client).

File: syft/workers/tfe.py (hunk 7 above)
Comment (mortendahl) [reply syft/workers/tfe.py]: the script launched below needs it

File: syft/frameworks/keras/model/sequential.py (hunk 8 above)
Comment (mortendahl) [reply syft/frameworks/keras/model/sequential.py]: not sure, by the same logic we could also have used `share_to_workers` and `serve_on_workers` yet `workers` seem implicit in those actions

Comment (mortendahl) [reply syft/frameworks/keras/model/sequential.py]: @iamtrask what's your take on this naming? what fits best with the terminology you guys are using elsewhere?

File: syft/frameworks/keras/model/sequential.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mortendahl) [reply syft/frameworks/keras/model/sequential.py]: I think we should use `model.get_weights` eventually -- it'll be useful for loading in from e.g. a file as well

File: examples/experimental/tfe_integration_demo/Step1_Public_Training.ipynb (hunk 9 above)
Comment (jvmancuso) [reply examples/experimental/tfe_integration_demo/Step1_Public_Training.ipynb]: actually, there's an issue in TFE blocking this currently (https://github.com/tf-encrypted/tf-encrypted/issues/499#issuecomment-495412270). suggest keeping as is until that issue is resolved

File: examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb [hunk 14]
```diff
@@ -0,0 +1,312 @@
+{
```
Comment (iamtrask) [inline examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb]: <p>- extra comma "...define your Syft Keras model, like we did..."</p><p><br></p><p>- "share" doesn't tell us what is happening in the description (aside from plugging TFE). It should be talking about protocols (SPDZ, SecureNN, etc.), not libraries.</p><p><br></p><p>- "serve" also doesn't say what's happening. It says "what you'll be ready for" but it needs to say "this opens up an endpoint on X ports/hostnames... etc."</p><p><br></p><p>- "shutdown_workers" - this is the first time we mention a "worker". This should be mentioned earlier.</p><p><br></p><p>Some of these can be addressed by moving the bottom paragraph up to the top and then referencing it.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (iamtrask) [inline examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb]: <p>"For this demo will send input data with the shape of (1, 28, 28, 1)." is not grammatical.</p><p><br></p><p>"We also return the logit instead of softmax because this operation has not yet been implemented in TF Encrypted." could use a bit more explanation but I generally get it.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (iamtrask) [inline examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb]: <p>Explain the difference between manual and automatic use. </p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (iamtrask) [inline examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb]: <p>num_steps is a poorly named variable. I get what it means but the first time user probably won't. We need to specify that the server is expecting to serve only a limited number of predictions.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (iamtrask) [inline examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb]: <p>We should change "Served" to be much more informative. Ideally "produced encrypted prediction to &lt;client name&gt;" or at least a list of the servers involved.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: examples/tutorials/Part 11c - Secure Classification with Syft Keras and TFE: Private Prediction Client.ipynb [hunk 15]
```diff
@@ -0,0 +1,206 @@
+{
```
Comment (iamtrask) [inline examples/tutorials/Part 11c - Secure Classification with Syft Keras and TFE: Private Prediction Client.ipynb]: <p>grammatical error</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (iamtrask) [inline examples/tutorials/Part 11c - Secure Classification with Syft Keras and TFE: Private Prediction Client.ipynb]: <p>Before querying the model, you have to connect to it. To do so, you first create a client. Then, you define the exact same three TFEWorkers that we defined in the last notebook (<code>alice</code>,&nbsp;<code>bob</code>, and&nbsp;<code>carol</code>). Finally call&nbsp;<code>connect_to_model</code>.</p><p><br></p><p>we also need more explanation about what "connect_to_model" is doing.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (iamtrask) [inline examples/tutorials/Part 11c - Secure Classification with Syft Keras and TFE: Private Prediction Client.ipynb]: <p>Replace with:</p><p><br></p><p>This is great. You are able to classify these three images correctly! But what's special about these predictions is that you haven't revealed any private information to get this service. The model host never saw your input data or your predictions, and you never downloaded the model. You were able to get private predictions on encrypted data with an encrypted model!</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2170/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb (hunk 14 above)
Comment (jvmancuso) [reply examples/tutorials/Part 11b - Secure Classification with Syft Keras and TFE: Secure Model Serving.ipynb]: <p>changed, couldn't get a client name unfortunately but it's more informative now</p>
